torch/csrc/distributed/autograd/context/context.h

```diff
@@ -122,7 +125,7 @@ class TORCH_API DistAutogradContext {
   std::vector<std::shared_ptr<rpc::FutureMessage>> outStandingRpcs_;
 
   // Lock to protect concurrent modification of the context.
-  mutable std::mutex lock_;
+  mutable std::recursive_mutex lock_;
```

Comment:
I was taking a second look and I was getting bad vibes about the recursive mutex, so I took a closer look here. I'm not convinced by this explanation to Alban's comment. A few things:

1. The problem with recursive mutexes is a lot of things just don't "work correctly" when you do a recursive mutex. For example, if you do a condition variable wait on a mutex that has been recursively locked, you will still maintain the lock in the outer scope (likely leading to a deadlock). I don't think the code as you've written here commits this mistake, but there are just a lot more things that can go wrong when a recursive mutex is involved and so it just makes it harder to code review the code in the future.

2. The most common reason to introduce a recursive mutex is because you can't conveniently rearrange your code so that you never call locking functions from inside locking functions; e.g., maybe there is a pretty nested call chain and you don't want to created an unlocked version of every function. But that's not the case here: you're directly calling `set_exception` which is the only thing which recursively takes out the lock; you then observe that it might call `resetGraphTask` eventually but we see:

```
void GraphTask::set_exception(
    std::exception& e,
    const std::shared_ptr<Node>& fn) {
  std::unique_lock<std::mutex> lock(mutex_);
  if (!has_error_.load()) {
    if (AnomalyMode::is_enabled() && fn) {
      fn->metadata()->print_stack();
    }
    has_error_ = true;
    // Careful: setting the future_result_ can trigger DistAutogradContext to
    // resetGraphTask(), sometimes deleting this underlying GraphTask.
    // Don't touch *this after setError() below, and release the lock early, to
    // avoid unlocking this->mutex_ after setting the future.
    std::shared_ptr<FutureVariableList> future_result = future_result_;
    lock.unlock();
    if (!future_result->completed()) {
      future_result->setError(e.what());
    } else {
      TORCH_INTERNAL_ASSERT(future_result->hasError());
    }
  }
}
```

This avoids the recursive mutex call, AND it is good logic: at the point in time you are calling setError on the future result, you don't need the lock on the GraphTask, so you ought to release it before calling arbitrary user code.

More concretely, I don't see why you can't refactor `set_exception` into a version that takes in the lock that it should release before calling the future callback, and then call that direct version with the parent lock. Then, as far as I can tell, you don't need a recursive mutex at all. This is a much more local change.